Doc/includes/noddy3.c

```diff
@@ -59,167 +54,126 @@ Noddy_init(Noddy *self, PyObject *args, PyObject *kwds)
         self->first = first;
         Py_DECREF(tmp);
     }
-
     if (last) {
         tmp = self->last;
         Py_INCREF(last);
         self->last = last;
         Py_DECREF(tmp);
     }
-
     return 0;
 }
 
 static PyMemberDef Noddy_members[] = {
-    {"number", T_INT, offsetof(Noddy, number), 0,
+    {"number", T_INT, offsetof(NoddyObject, number), 0,
      "noddy number"},
     {NULL}  /* Sentinel */
 };
 
 static PyObject *
-Noddy_getfirst(Noddy *self, void *closure)
+Noddy_getfirst(NoddyObject *self, void *closure)
 {
     Py_INCREF(self->first);
     return self->first;
 }
 
 static int
-Noddy_setfirst(Noddy *self, PyObject *value, void *closure)
+Noddy_setfirst(NoddyObject *self, PyObject *value, void *closure)
 {
     if (value == NULL) {
         PyErr_SetString(PyExc_TypeError, "Cannot delete the first attribute");
         return -1;
     }
-
-    if (! PyUnicode_Check(value)) {
+    if (!PyUnicode_Check(value)) {
         PyErr_SetString(PyExc_TypeError,
                         "The first attribute value must be a string");
         return -1;
     }
-
-    Py_DECREF(self->first);
     Py_INCREF(value);
+    Py_CLEAR(self->first);
```

Comment:
You should use the dance with `tmp` here. `Py_CLEAR(self->first)` calls the destructor. The `first` attribute can be set directly in the destructor or it can be set in other thread when the GIL will be released in the destructor. `self->first = value` in the following line will rewrite the new value of the `first` attribute with leaking a reference.

`Py_SETREF()` could help here, but it is not in a public API.

Reply:
`Py_CLEAR` should first transfer the pointer to a temporary before decref'ing it, so I don't see where the problem is.

Reply:
The problem is virtually the same as in `__init__`.

* Thread A calls `Noddy_setfirst` with the argument *valueA*, runs to `Py_CLEAR(self->first)` in which sets `tmp = self->first; self->first = NULL` and executes `Py_XDECREF(tmp)` which calls the destructor and releases the GIL. This causes switching to other thread.
* Thread B calls `Noddy_setfirst`, and sets `self->first` to a new value *valueB* (it was set to NULL in thread A).
* Thread A continues execution. The next line is `self->first = valueA`. It is expected that the value of `self->first` is NULL (as set by `Py_CLEAR`), but actually it is *valueB*. After setting `self->first = valueA` *valueB* will be leaked.

The same race condition is possible in a single-thread case if the destructor of the old attribute value calls `Noddy_setfirst`.